src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs

```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
             {
-                throw new ArgumentException(Resources.Settings_FileName_Cannot_Be_A_Path, nameof(fileName));
+                if (section.TryUpdateChildItem(item, isBatchOperation))
+                {
+                    return true;
+                }
             }
 
-            Root = root;
-            FileName = fileName;
-            XDocument config = null;
-            ExecuteSynchronized(() => config = XmlUtility.GetOrCreateDocument(CreateDefaultConfig(), ConfigFilePath));
-            ConfigXDocument = config;
-            IsMachineWideSettings = isMachineWideSettings;
-            CheckConfigRoot();
+            var outputSettingsFile = GetOutputSettingsFileForSection(sectionName);
+
+            // Set is an add
+            return outputSettingsFile != null && outputSettingsFile.RootElement.AddItemInSection(sectionName, item.Copy(), isBatchOperation);
         }
 
         public event EventHandler SettingsChanged = delegate { };
 
-        public IEnumerable<ISettings> Priority
+        public Settings(string root)
+            : this(new SettingsFile(root)) { }
+
+        public Settings(string root, string fileName)
+            : this(new SettingsFile(root, fileName)) { }
+
+        public Settings(string root, string fileName, bool isMachineWide)
+            : this(new SettingsFile(root, fileName, isMachineWide)) { }
+
+        private Settings(SettingsFile settingsHead)
+        {
+            _settingsHead = settingsHead;
+
+            var curr = _settingsHead;
+            while (curr != null)
+            {
+                curr.SettingsChanged += (_, __) => { RefreshSettingValues(); };
+                curr = curr.Next;
+            }
+
+            MergeSettingFilesValues();
+        }
+
+        private void RefreshSettingValues()
+        {
+            MergeSettingFilesValues();
+            SettingsChanged?.Invoke(this, EventArgs.Empty);
+        }
+
+
+        private void MergeSettingFilesValues()
+        {
+            var computedSections = new Dictionary<string, SettingsSection>();
+
+            var curr = _settingsHead;
+            while (curr != null)
+            {
+                curr.RootElement.MergeSectionsInto(computedSections);
+                curr = curr.Next;
+            }
+
+            _computedSections = computedSections;
+        }
+
+        private ISettingsFile GetOutputSettingsFileForSection(string sectionName)
+        {
+            // Search for the furthest from the user that can be written
+            // to that is not clearing the ones before it on the hierarchy
+            var writteableSettingsFiles = Priority.Where(f => !f.IsMachineWide);
+
+            var clearedSections = writteableSettingsFiles.Select(s => s.RootElement.GetSection(sectionName)).Where(s => s != null && s.HasClearChild());
+            if (clearedSections.Any())
+            {
+                return clearedSections.Last().Origin;
```

Comment:
It seems your `SettingSection` is doing all the magic... :) But it's really hard to read... not sure if this was the existing code but anything we can do to improve readability?